app/javascript/packages/document-capture/components/in-person-location-step.tsx

```diff
@@ -1,14 +1,190 @@
+      throw error;
+    }),
+  );
+  return response;
+};
+
+const formatLocation = (postOffices: PostOffice[]) => {
+  const formattedLocations = [] as FormattedLocation[];
+  postOffices.forEach((po: PostOffice, index) => {
+    const location = {
+      formattedCityStateZip: `${po.city}, ${po.state}, ${po.zip_code_5}-${po.zip_code_4}`,
+      id: index,
+      name: po.name,
+      phone: po.phone,
+      saturdayHours: po.saturday_hours,
+      streetAddress: po.address,
+      sundayHours: po.sunday_hours,
+      weekdayHours: po.weekday_hours,
+    } as FormattedLocation;
+    formattedLocations.push(location);
+  });
+  return formattedLocations;
+};
+
+const snakeCase = (value: string) =>
+  value
+    .split(/(?=[A-Z])/)
+    .join('_')
+    .toLowerCase();
+
+// snake case the keys of the location
+const prepToSend = (location: object) => {
+  const sendObject = {};
+  Object.keys(location).forEach((key) => {
+    sendObject[snakeCase(key)] = location[key];
+  });
+  return sendObject;
+};
 
 function InPersonLocationStep() {
   const { t } = useI18n();
+  const [locationData, setLocationData] = useState([] as FormattedLocation[]);
+  const [inProgress, setInProgress] = useState(false);
+  const [autoSubmit, setAutoSubmit] = useState(false);
+  const [isLoadingComplete, setIsLoadingComplete] = useState(false);
+
+  // ref allows us to avoid a memory leak
+  const mountedRef = useRef(false);
+
+  useEffect(() => {
+    mountedRef.current = true;
+    return () => {
+      mountedRef.current = false;
+    };
+  }, []);
+
+  // useCallBack here prevents unnecessary rerenders due to changing function identity
+  const handleLocationSelect = useCallback(
+    async (e: any, id: number) => {
+      if (autoSubmit) {
+        return;
+      }
+      // prevent navigation from continuing
+      e.preventDefault();
+      if (inProgress) {
+        return;
+      }
+      const selected = prepToSend(locationData[id]);
+      const headers = { 'Content-Type': 'application/json' };
+      const meta: HTMLMetaElement | null = document.querySelector('meta[name="csrf-token"]');
+      const csrf = meta?.content;
+      if (csrf) {
+        headers['X-CSRF-Token'] = csrf;
+      }
+      setInProgress(true);
+      await fetch(locationUrl, {
+        method: 'PUT',
+        body: JSON.stringify(selected),
+        headers,
+      })
+        .then(() => {
+          if (!mountedRef.current) {
+            return;
+          }
+          setAutoSubmit(true);
+          setImmediate(() => {
+            // continue with navigation
+            e.target.click();
+            // allow process to be re-triggered in case submission did not work as expected
+            setAutoSubmit(false);
+          });
+        })
+        .finally(() => {
+          if (!mountedRef.current) {
+            return;
+          }
+          setInProgress(false);
+        });
+    },
+    [locationData, inProgress],
+  );
+
+  useEffect(() => {
+    let mounted = true;
```

Comment:
(Non-blocking, suggestion) There's a [`useIfStillMounted` helper React hook](https://github.com/18F/identity-idp/blob/main/app/javascript/packages/react-hooks/use-if-still-mounted.ts) which seems like it might help for what's being done here? Alternatively, for this kind of thing, we may even consider pulling in a library like [`swr`](https://www.npmjs.com/package/swr) or [`react-query`](https://tanstack.com/query/v4) to help with the asynchronous data fetching.

This could also apply for `mountedRef` and the `cancelled` variable in `InPersonPrepareStep`.